# Context 

Mappability can be challenging in repetitive regions when analyzing NGS-derived data. Local **duplication of genomic** regions hampers the alignment of reads in those parts of the genome, since it **prevents unambiguously** **alingning reads**. Therefore, those reads are usually assigned a random location (among the possible ones) and a low _mapping quality_. Since reads are assigned a random location, **mutations** **are usually either** **missed** (mutated reads are dispersed among all possible alignment regions) or, even worse, **called multiple times** leading to the rise of false positives (enough reads to call a variant in multiple alignment regions). For this reason, **low mapping quality regions are often ignored** in variant calling pipelines.

__Armadillo__ is a somatic variant calling pipeline that aims at repetitive regions for tumor-control paired samples, in order to recover mutations that are systematically missed. In order to avoid both false positives and false negatives, it first __finds all candidate regions__ for a repetitive region of interest (RROI). Then, it extracts all reads coming from all regions and aligns them at the RROI sequence only instead of the whole genome, forcing all reads to be aligned together. This way, dilution of variant-supporting reads is prevented, as well as the false positives caused by mistakenly call a mutation at multiple copies. 
Finally, the variant calling step is carried out. After usual __heuristic filters__, we use __context-specific filters__ to prevent false positives related the stacking step and, lastly, run a __CNN model to classify the mutations__.

## In-depth review of Armadillo

As stated before, calling mutations at repetitive region can be risky, since we'll probably find false positives. In order to prevent this, Armadillo will:

1. **Retrieve all copies of a repetitive region of interest** (RROI). We just need to pass a list of regions of interest and it will find the other copies of each region we are interested in.
2. **Extract the reads** **aligned at each locus** and align them against the sequence of our RROI (instead of the whole genome). This way, **the alignment will be unambiguous** since all reads that could be aligned at each copy will be forced to be aligned at a single locus.
3. Apply **heuristic filters** to select candidate variants (at least 3 variant-supportive reads in tumor, less than 3 variant-supportive reads in control sample to remove germinal variants, minimum coverage...)
4. Apply **specific filters**. The step 2 assures that all possible reads are being used, increasing sensitivity, but it also leads to an increase of false positives (artifacts arise). It checks that nearby variants (**context**) **match** in all mutated reads, it tries to **find non-mutated reads that carry the same context** in the control sample..., so false positives are controlled. 
5. Perform **MonteCarlo simulations** to compute the chance of our mutation being somatic or germinal. 
6. Run a **CNN model** trained with curated mutations from our paper that takes into account the context for calling the mutation.
7. Make sure that the same mutation isn't returned multiple times. If a gene has two copies (gene1, gene2) and both gene1 and gene2 have been used as input, only the one will be returned in the final file.

More information on the pipeline can be retrieved on our [GitHub repository](https://github.com/xa-lab/armadillo) or in our paper (unpublished, 2022).

# Inputs

## Required inputs

1. `name`: Experiment name
2. `control_genome`: Control sample genome
3. `tumor_genome`: Tumor sample genome
4. `reference_blat`: 2bit file to use with blat (hg19, hg38 and mmu files available at [UCSC](https://hgdownload.soe.ucsc.edu/downloads.html)) 

## Optional inputs

1. `controlIndex` and `tumorIndex`: BAM index generated with `samtools`. Recommended for shorter runs
2. `armadillo_data`: armadillo data-prep output compressed as tar.gz 
4. `rois_list`: List of ROIs to analyze instead of the whole list in armadillo_data (BED format or chr:start-end format accepted). Header not needed.
5. `genome_ref`: Reference genome fasta
6. `genome_ref_index`: Reference genome index
7. `return_bams`: Return minibams generated by armadillo
8. `return_dataprep`: Return data-prep results (useful if same analysis will be done on a batch of samples)
9. `return_discarded`: Return discarded variants

> __Note__: Either `armadillo_data` or `rois_list` is needed. Also, `genome_ref` is required if `armadillo_data` is not provided.

## Additional parameters

The `params` argument is an array of parameters to adjust some of the filters applied downstream. It includes:

|     Parameter     |                             Description                            | Default |
|:-----------------:|:------------------------------------------------------------------:|:-------:|
|  control_coverage |                   Coverage of the control sample                   |    30   |
|   tumor_coverage  |                    Coverage of the tumor sample                    |    30   |
| control_threshold | Maximum alternative-allele supportive reads allowed in the control |    3    |
|  tumor_threshold  |      Minimum coverage required for a variant to be considered      |    6    |
|    base_quality   |         Minimum base quality required for the tumor genome         |    30   |
|        mapq       |            Minimum MapQ for reads after being collapsed            |    30   |
|      GCcutoff     |                         Maximum GC% allowed                        |    80   |

Also, `dataprep_params` allows tuning some arguments when doing a brand new dataprep file:

| Parameter |                                Description                                | Default |
|:---------:|:-------------------------------------------------------------------------:|:-------:|
|  identity |          Minimum identity required to consider a region as a copy         |    95   |
|  lendiff  | Maximum length difference allowed between each hit and the input sequence |    15   |
|    mlen   |                  Minimum length (bp) allowed to each gene                 |   100   |

# Outputs

1. `vcf`: VCF with the mutations detected by the pipeline
2. `dataprep`: data-prep output compressed as tar.gz (_optional_, only if `return_dataprep=true`)
3. `tumor_minibam` and `control_minibam`: minibams generated by the pipeline (_optional_, only if `return_bams=true`)
4. `tumor_minibam_idx` and `control_minibam_idx`: minibam indexes (_optional_, only if `return_bams=true`)
5. `discarded variants`: log file with all discarded candidates (_optional_, only if `return_discarded=true`)

# Example 

1. Download a 2bit file valid with your aligned bams (same reference version): 
```bash
wget https://hgdownload.cse.ucsc.edu/goldenpath/hg38/bigZips/hg19.2bit
```

2. Copy bams, ROIs and 2bit file to batchX
```bash
bx cp tumor.bam control.bam hg19.2bit rois bx://test/
```

> A subset of rois from our paper is available in our [repo](https://github.com/xa-lab/armadillo/blob/master/lib/armadillo_data/rois).

3. Submit job

```bash
bx submit -v=10 -m=16000 uniovi@labxa/armadillo:1.0.0 \
'{
  "name": "test",
  "control_genome": "/test/control.bam",
  "tumor_genome": "/test/tumor.bam",
  "reference_blat": "/test/hg19.2bit",
  "rois": "/test/rois",
}'
```

# FAQ

- **What is a BAM? What is the index?**

A BAM file is the output we get after aligning the reads against a reference genome. This step is absolutely required, since this type of files report not only the sequences but also information about the location of the reads in the reference genome and mappability (mapQ). The index is not required, but __recommended__. The indexes allow the pipeline to retrieve information of a region directly, instead of reading the whole file (very time-consuming). If not provided, the pipeline will index them automatically, but it will take some minutes.

If your sequences are in FASTQ or FASTA format, we recommend to align them with [BWA](https://platform.batchx.io/batchx/images/bioinformatics%2Fbwa%2Fmem/2.0.2). Also, the BAM indexes can be created with [SAMtools index](https://platform.batchx.io/batchx/images/bioinformatics%2Fsamtools%2Findex/1.0.0#documentation).

- **How do I get a file of regions of interest?**

Armadillo doesn't analyze the whole genome as regular variant callers. It relies in a preset of regions that may or may not be repetitive, but it actually needs to be told where to find the mutations. Then, it will check which of those regions are repetitive, and will call mutations exclusively in those. A simple tabulated file with chromosome, start and end is enough. It also accepts chr:start-end format.  

- **What is a .2bit file?**

It's a file for storing DNA sequences in a compact randomly-accesible way. It's used by `gfClient` for performing BLAT-like queries. It allows to quickly align a read or a sequence in the genome. More information can be found in the [UCSC](http://genome.ucsc.edu/FAQ/FAQformat.html#format7). Also, 2bit and fasta files of common species (human, mice...) can be downloaded [here](https://hgdownload.soe.ucsc.edu/downloads.html). 

- **Does the final VCF need any postprocessing?**

No, it does not. Nonetheless, if you are using a batch of samples, using a panel of normals can be helpful to reduce the noise. Panel of normals are collections of variants that appear in multiple healthy samples. If a somatic mutation exists in those panels, it's not expected to be a cancer driver. Also, it can help removing some unexpected false positives. In order to do so with our data, use `return_discarded: true` in your input files. This way, a list of all discarded mutations will be returned. Then, you can merge all `discarded_variant.log` files and keep all variants with the flag `germline_change`. This way, a list of variants found in the healthy sample will be returned. After that, count the occurrences of each mutation. Short example in linux:
```
cd $my_results # Directory where all armadillo results are saved
cat */discarded_variants.log | grep germline_change | cut --complement  -f3 | sort | uniq -c | awk '{first = $1; $1=""; print $0, first}' > mypanel.txt
```
Now, this database can be used to know if a variant appears recurrently in healthy samples.


# Tools version

* [BWA](http://bio-bwa.sourceforge.net/) (built with v.0.7.17)
* [Samtools](https://www.htslib.org/doc/1.9/samtools.html) (built with v.1.9)
* [gfClient](https://genome.ucsc.edu/goldenPath/help/blatSpec.html#gfClientUsage)  (built with v.35)
* [Python3](https://www.python.org) (built with v.3.7)

# Authors

* **Pablo Bousquets Muñoz** - bousquetspablo@uniovi.es
* **Ander Díaz Navarro**
* **Xose Antón Suárez Puente**
